tiny_dnn/layers/recurrent_cell_layer.h

```diff
@@ -0,0 +1,174 @@
+    return in_shape()[i].height_;
+  }
+
+  std::vector<index3d<serial_size_t>> in_shape() const override {
+    if (params_.has_bias_) {
+      return {
+        index3d<serial_size_t>(params_.in_size_, 1, 1),   // x
+        index3d<serial_size_t>(params_.out_size_, 1, 1),  // h(t-1)
+        index3d<serial_size_t>(params_.in_size_, params_.out_size_, 1),   // U
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1),  // W
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1),  // V
+        index3d<serial_size_t>(params_.out_size_, 1, 1),                  // b
+        index3d<serial_size_t>(params_.out_size_, 1, 1)};                 // c
+    } else {
+      return {
+        index3d<serial_size_t>(params_.in_size_, 1, 1),   // x
+        index3d<serial_size_t>(params_.out_size_, 1, 1),  // h(t-1)
+        index3d<serial_size_t>(params_.in_size_, params_.out_size_, 1),    // U
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1),   // W
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1)};  // V
+    }
+  }
+
+  std::vector<index3d<serial_size_t>> out_shape() const override {
+    return {index3d<serial_size_t>(params_.out_size_, 1, 1),
+            index3d<serial_size_t>(params_.out_size_, 1, 1)};  // h(t)
+  }
+
+  void forward_propagation(const std::vector<tensor_t *> &in_data,
+                           std::vector<tensor_t *> &out_data) override {
+    // forward recurrent op context
+    auto ctx = OpKernelContext(in_data, out_data);
+    ctx.setParallelize(layer::parallelize());
+    ctx.setEngine(layer::engine());
+
+    // launch recurrent kernel
+    kernel_fwd_->compute(ctx);
+  }
+
+  void back_propagation(const std::vector<tensor_t *> &in_data,
+                        const std::vector<tensor_t *> &out_data,
+                        std::vector<tensor_t *> &out_grad,
+                        std::vector<tensor_t *> &in_grad) override {
+    // backward recurrent op context
+    auto ctx = OpKernelContext(in_data, out_data, out_grad, in_grad);
+    ctx.setParallelize(layer::parallelize());
+    ctx.setEngine(layer::engine());
+
+    // launch recurrent kernel
+    kernel_back_->compute(ctx);
+  }
+
+  std::string layer_type() const override { return "recurrent"; }
+
+#ifndef CNN_NO_SERIALIZATION
+  friend struct serialization_buddy;
+#endif
+
+ protected:
+  void set_params(const serial_size_t in_size,
+                  const serial_size_t out_size,
+                  bool has_bias) {
+    params_.in_size_    = in_size;
+    params_.out_size_   = out_size;
+    params_.has_bias_   = has_bias;
+    params_.activation_ = new activation_type;
+  }
+
+  void init_backend(backend_t backend_type) {
+    core::OpKernelConstruction ctx =
+      core::OpKernelConstruction(layer::device(), &params_);
+
+    if (backend_type == backend_t::internal) {
+      kernel_fwd_.reset(new RecurrentCellOp(ctx));
```

Comment:
Hi thanks! I have fixed it.

Indeed, there is a lot of repetition with the backends. I wanted to simplify it globally with this PR but I think it might be better to do a new PR removing all the backend checks until the ops.

Having a default() somewhere seems correct to me, but it feels forced right now.